sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,392 @@
+
+|      | Solution                                        | Description                                                  | Pros                                                         | Cons                                                         |
+| ---- | ----------------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| 1    | Keep as is                                      |                                                              | - Doesn't break current user apps<br />- Seems to be the direction the industry is going | - 92 lines and 9 params so far makes it unweidy<br />- Constructor is overloaded, handling multiple responsibilities<br />- Doesn't take advantage of OOP |
+| 2    | Config/Creds objects for *each* supported flow  | ConfigurationFromFile()  - .conjurrc<br />ConfigurationFromEnv() - ENV vars<br />CredentialsFromFile() - netrc<br />CredentialsFromKeystore() - keystore | - Explicit, the flow the user chooses is very clear<br />- Security is up to user's the discretion | - Conflicting experience with CLI. The CLI performs a discovery on Config/Creds retrieval type<br />- Requires changes when a user wants to choose another flow |
+| 3    | Generic Config/Creds object                     | Configuration()<br />Credentials()<br />and we will discover where the connection details live based on environment's context | - Easy maintainability. No change in user's code as we build new ways to fetch config and creds<br />- Aligns with CLI UX of Config/Cred discovery | - Implicit and the flow we choose is not clear               |
+|      | A single Config/Cred object with defining param | Configuration(action=file) - .conjurrc<br />Credential(action=keystore) | - Explicit<br />- Minimal code changes should the user choose a different flow<br />- Strong drive in Python to be as explicit as possible | - Requires changes in code if user wants to adopt another retrieval method<br />- Conflicting experience with CLI. The CLI preforms a discovery on Config/Creds retrieval type |
+
+**Decision:** Option #4. 
+
+The flow for configuration and credential will be determined *explicitly* by enabling the user to pass in the flow of their choosing via a parameter to the object during instantiation. 
+
+This functionality will be implemented using the [Factory pattern](https://refactoring.guru/design-patterns/factory-method). Each factory-generated class will have its own implementation and validation mechanism. This will make the implementation easy to maintain as this makes it easy to swap algorithms used inside an object during runtime. That way, we can isolate the implementation details from the *Client*. If we choose to implement a new strategy, for example fetching values from environment values, this will not require a change in the Client.
```

Comment:
Regarding the Client structure I suggested something like that:
` class Client:

    def __init__(self, credentials, connection_data, **kwargs):
        self.credentials = credentials
        self.connection_data = connection_data
        self._variable = None
        self._policy = None`
        `    @property
    def variable(self):
        if self._variable is None:
            self._variable = Variable(self.credentials, self.connection_data)
        return self._variable`
        `class Variable:
    def __init__(self, credentials, connection_data):
        self.credentials = credentials
        self.connection_data = connection_data

    def get_one(self, variable_id):
        pass

    def get_many(self, ids):
        asyncio.run(self.get_many_async())

    async def get_many_async(self):
        pass`
        
The way we going to use this interface is by running 
        `    cred = CredentialsData()
    connectionData = ConjurrcData()
    client = Client(cred, connectionData)
    client.variable.get_many([1, 2, 3])`